### PR TITLE
ViewPos

### PR DESCRIPTION
**Pull request by MajorCooke**
Added ViewPos vector3, which offsets the actor view.by the specified amount. Primarily used by cameras and players.

- VIEWPOSNOANGLES: Angles (yaw/pitch/roll) are not taken into account, but the position is still relative.
- ABSVIEWPOS: Absolute position of the view in world coordinates. When used, portal interpolation is **NOT** performed, which is left to the modder to handle instead.

Currently, the only known problem is the inability to turn off the player sprite when the view point and the player are separated via portals, but should otherwise be ready for implementation.

### CODE REVIEW

File: src/playsim/actor.h [hunk 1]
```diff
@@ -413,6 +413,8 @@ enum ActorFlag8
 	MF8_RECREATELIGHTS	= 0x00000100,	// Internal flag that signifies that the light attachments need to be recreated at the
 	MF8_STOPRAILS		= 0x00000200,	// [MC] Prevent rails from going further if an actor has this flag.
 	MF8_ABSVIEWANGLES	= 0x00000400,	// [MC] By default view angle/pitch/roll is an offset. This will make it absolute instead.
+	MF8_VIEWPOSNOANGLES	= 0x00000800,	// [MC] View position takes no angles into account
```
Comment (MekBoss) [inline src/playsim/actor.h:416]: Isnt having VIEWPOSIGNOREYAW(PITCH/ROLL) here would be more suitable? I mean three flags for each camera rotation axis.

**Review by alexey-lysiuk — COMMENTED** [review]
Every change to `AActor` has exactly the same problem. The bloated class becomes even more bloated.
This doesn't mean that this PR is wrong or useless. Please note, however, that `sizeof(AActor)` is already 1424 bytes on 64-bit platforms.
An addition of a new member(s) by itself isn't responsible for any apparent performance issues. Although, thanks to a number of actors in modern maps and mods, we cannot simply ignore such influence.

**Comment by MajorCooke** [comment]
I'd rather just introduce one flag over three. Less bloat.

**Comment by alexey-lysiuk** [comment]
My comment was not about flags. It was about a new `AActor::ViewPos` member.

**Comment by MajorCooke** [comment]
I was talking to MekBoss.

**Comment by MajorCooke** [comment]
But yes Alex, I am aware. However, whose going to be the one to introduce float vectors? It's not going to be me.

We've had this discussion before on several other PRs already.

What we really need is some way to separate a bunch of stuff from an actor into subclasses.

**Comment by alexey-lysiuk** [comment]
Subclasses won't fix anything in common case as we need to know the exact type at the place of construction.
There are no hints on which members will be required and which can be skipped.
It could help with only if we can be 100% sure that the given actor class won't change its properties.

In large engines, this problem is partially solved by entity-component or entity-component-system framework.
Entity is the same as actor in GZDoom. The idea is to have it as light as possible, i.e. with a very few members like position (global matrix usually), some common flags, and list of components.
All logic except basic entity management is encapsulated in components.
In our case, this means that we need to add/remove components on demand, when particular feature is used first for the given actor, or, in general, when related property has been modified.

In theory, this may work. In practice, no one knows. It's not the change that can be done in one evening.

**Comment by MajorCooke** [comment]
Hmmm. I'm curious if such a thing were implemented, would it also include a function for checking if a variable is present or a certain preset is loaded with the actor? In a way it seems similar to mixins but without the limits of translation units.

**Comment by alexey-lysiuk** [comment]
This has nothing to do with mixins, at least in a sense of compiled languages.
There is a list of components in every entity, and these components determine what particular entity can do.
With entity-component framework properly implemented, types of components added to the given entity specify its “capabilities”.

**Comment by MekBoss** [comment]
Ehhhhh, its not the right topic, but it would have much more sense to "extend capabilities" of thinker, rather than actor IMO.

Actors has been used for, how much, more than 10 years? Such global changes in its internal structure can made old mods incompatible and/or buggy and/or %insert bad thing name% with Gzdoom versions which would implement component system by altering behavior/structure of an actor class. Nothing that several hacks, crutches and bunch of duct tape cant fix, of course, but...

At the same time, thinker, after exposing to modders through Zscript, mostly was used as ticking thing as/for ACS replacement.
So adding to it something like `list/array/vector<components> capabilities;` and `foreach(component c in capabilities)` in its tick will not break old mods/code. At least I think so.

**Comment by MajorCooke** [comment]
Curious, mental: if you could take a guess, unless you know a good way components can go, do you think it would require a whole new redefinition of Actor or would Actor be modified to simply include all the current properties and functions it already has to maintain compatibility while a new base actor class is defined?

Something tells me that Actor would need rewriting from the ground up.

@MekBoss I did already suggest that in a way for a [SpriteParticle](https://forum.zdoom.org/viewtopic.php?f=15&t=69383) class. If that ever does come to light, it'd be a HUGE step in the right direction alone for speeding up the game.

**Comment by alexey-lysiuk** [comment]
The idea is to have lightweight actor class. So, all existing members of `AActor` cannot remain where they are by definition.
ZScript definition of `Actor` must remain unchanged. All handling of a new internal actor representation needs to be done on C++ side.
The most obvious problem is a direct access to a property that must be rerouted to a corresponding component. Taking into account that component may not exist yet (property is accessed for the first time), extra cost in such scenario can easily overweight the gain from a small actor class.

**Comment by MajorCooke** [comment]
I've decided to open up this commit and prepare it for QZDoom. The main issues I had previously were fixed, even if the player itself is drawn whenever the player and the view point are separated by portal. Aside from this minor visual detail, everything else is working fine.

**Comment by coelckers** [comment]
> The idea is to have lightweight actor class. So, all existing members of `AActor` cannot remain where they are by definition.
> ZScript definition of `Actor` must remain unchanged. All handling of a new internal actor representation needs to be done on C++ side.
> The most obvious problem is a direct access to a property that must be rerouted to a corresponding component. Taking into account that component may not exist yet (property is accessed for the first time), extra cost in such scenario can easily overweight the gain from a small actor class.


To elaborate on this - every single property that needs to be checked by the renderer cannot be in a component it needs to be in the core entity, because it needs to be checked anyway. Doom originally solved these issues by making several properties unchangeable meta properties that are held in a separate structure.

Overall I cannot stress this enough: The main problem of the sprite rendering code is not feature bloat as-is but caching behavior. It is a clear sign for cache issues if it doesn't really matter how I optimize the code - the end result in time spent is always the same and always consists of one huge time spike when the renderer first accesses the actor.
The same pattern can be observed in the clipper as well when reading from the linedefs. 
This is also the main reason why PrBoom is faster. Despite everything that got added over time, our renderer is far better optimized, but it really does not help if niche features get constantly added to the core structures and cause poorer cache performance. The next round of optimizations will most definitely have to be to cut down on their size - offload rarely used features to child structs and determine via flag if they are needed at all. For example, all that gradient lighting or colorization stuff for walls and flats requires a lot of fields in the structs but is very rarely used, so if they could be toggled with a single bit and offloaded to the extsector type, a lot would be gained - just like flagging the existing features in there could possibly avoid unneeded cache misses.

This entire thing is a textbook example of such a feature - few people will ever need it, but everybody will suffer its downsides.

**Comment by Doom2fan** [comment]
> This entire thing is a textbook example of such a feature - few people will ever need it, but everybody will suffer its downsides.

I wouldn't go that far. This feature (along with the view angles one) would be fairly useful for player behaviour.
But that also poses a question - Is there any need for this to be in the Actor class in the first place? Most uses of it would be on players, so it could easily be added to the PlayerInfo struct instead, and cause no problems with cache misses if structured correctly.
For uses in non-player actors (which would be far more rare), you could just as well spawn an additional "camera" actor that copies the original actor's data, with the offsets applied. They don't support anything like PSprites, so doing this wouldn't cause issues like it would with players.

**Comment by MajorCooke** [comment]
I could do that if needed. But at the same time should I do the same thing with ViewAngles?

My main issue I was having there when I originally was implementing this was dealing with camera actors that are outside of it. I couldn't find a proper way to reset it or record what actors have certain offsets.

**Comment by MajorCooke** [comment]
Originally I tried putting it on a playerinfo struct but unfortunately, figuring out how certain actors wanted to record this information was a big problem. So I think what Graf suggested might be better.

Offloading into child structs sounds like it'd be the best.

**Comment by MajorCooke** [comment]
Once @coelckers performs optimization on everything, I'll follow suit. I just wasn't sure how best to handle this is all. I get the idea of putting it in structs and having a flag enable the execution, but I want to make sure I follow how he wants it done best.

**Comment by caligari87** [comment]
Sorry for the bump. I'm curious if there's any progress on this pull request? I've started work on a project that makes significant use of the related ViewAngles feature and it's been extraordinarily useful thus far. I'd love to be able to go further with it.

**Comment by MajorCooke** [comment]
Lets discuss it on the [forums](https://forum.zdoom.org/viewtopic.php?f=59&t=69780&p=1207191#p1207191). Hopefully we can attract more attention and opinions on how to handle this better.

**Comment by MajorCooke** [comment]
Closing this because I've a [newer and better branch. ](https://github.com/coelckers/gzdoom/pull/1517)

**Pull request closed without merging**